File: video_28_Longest_Palindromic_Subsequence.java

class video_28_Longest_Palindromic_Subsequence{
    public static String longestPalin(String str1){
        // code here
        int n = str1.length();
        String str2 = "";
        for(int i = n-1; i>=0; i--)
        str2+=str1.charAt(i);

        // System.out.println(str1);
        // System.out.println(str2);

        int maxi = 0;

        int dp[][] = new int[n+1][n+1];
        for(int i = 0; i<=n; i++){   
            dp[i][0] = 0;
            dp[0][i] = 0;
        }
        int r = 0;
        int c = 0;

        for(int i = 1; i<=n; i++){
            int ans = 0;
            for(int j = 1; j<=n; j++){
                if(str2.charAt(i-1)==str1.charAt(j-1)){
                    dp[i][j] = 1+dp[i-1][j-1];
                    ans = dp[i][j];
                    if(ans>=maxi){
                        r = i;
                        c = j;
                        maxi = ans;
                    }
                }
                else dp[i][j] = 0;
            }
        }

        String ans = "";
        int len = maxi;
        if(len==1){
           ans+=str1.charAt(0);
           return ans;
        }

        while(len>0){
             if(str2.charAt(r-1)==str1.charAt(c-1)){
                ans+=str2.charAt(r-1);
                len--;
             }
            r--;c--;
        }
        if(maxi==2){
            if(ans.charAt(0)!=ans.charAt(1)){
                String str = "";
                str+=str1.charAt(0);
                return str;
            }     
        }
        return ans;
    }

    public static void main(String[] args) {
        // Scanner sc = new Scanner (System.in);
        String str = "epzshkljugggekjdqzjenpevqgxiepjsrdzjazujllchhbfqmkimwzobiwybxduunfsksrsrtekmqdcyzjeeuhmsrqcozijipfioneeddpszrnavymmtatbdzqsoemuvnpppsuacbazuxmhecthlegrpunkdmbppweqtgjopa";
        // String str = "otafsngqvoijxuvqbztv";
        // String str = "ayaxzfbjbkrxiri";
        System.out.print(longestPalin(str));
    }
}

File: video_28_LongestPalindromicSubsequence.cpp
// #include<iostream>
// #include<bits/stdc++.h>
// using namespace std;

// int lcs(string str1, string str2){
//     int n = str1.size(), m = str2.size();
//     int ans = 0;
//     vector<vector<int>> dp(n+1, vector<int> (m+1, 0));

//     for(int i = 0; i<=n; i++) dp[i][0] = 0;
//     for(int j = 0; j<=m; j++) dp[0][j] = 0;

//     for(int i = 1; i<=n; i++){
//         for(int j = 1; j<=m; j++){
//             if(str1[i-1] == str2[j-1]){
//                 dp[i][j] = 1+dp[i-1][j-1];
//                 ans = max(ans, dp[i][j]);
//             }
//             else dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
//         }
//     }
//     // for(int i = 0; i<=n; i++){
//     //         for(int j = 0; j<=m; j++) cout<<" "<<dp[i][j];
//     //     cout<<endl;
//     // }
//     return ans;
// }
// int longestPalindromeSubsequence(string str1)
// {
//     // Write your code here.
//     string str2 = str1;
//     reverse(str2.begin(), str2.end());
//     return lcs(str1, str2); 
// }

// int main(){
//     string str;
//     cin>>str;
//     cout<<longestPalindromeSubsequence(str);
//     return 0;
// }
